### PR TITLE
`CpuWriteGpuReadBelt` for fast frame by frame memory transfers

### PR DESCRIPTION
**Pull request by Wumpf**
Rewrite of #594 
Introduces a convenient mechanism to allow writing directly to gpu readable memory which gpu then can copy into its internally gpu-only memory.

*Goal is faster (on less memcpy), safer (alignment guarantee) and easier (no need to create buffers manually) transfers.*

An unexpected ripple effect we got was that in order to have much higher memory use on Web, we need to limit the number of frames (technically: queue submissions) in flight. This makes it a lot less memory hungry as we accumlate a lot less per-frame data! Empirically, it looks a bit like wgpu handles resource cleanup a bit differently on native than on web which might just come down to the present-mode it chooses automatically. Given that the introduced limiter & `device.poll` makes sense regardless I haven't investigated this further.

Only used as a proof of concept in a single uniform buffer as well as point colors.

Nevertheless, some early perf numbers, need to repeat this test when we have it it more places:
(Acquired by selecting a small timeline range range and letting it loop, then opening profiler, waiting a bit, going to profiler, pause it, select a range of frames where the app was still in focus)
```
Nyud

DEV:
                            before   |   after (still in dev)
--------------------------------------------------------------
PointsBuilder::colors         2.280 |  2.281
PointCloudDrawData::new()     2.170 |  1.372
Overall                      12.715 | 11.762

RELEASE:
                            before   |   after (still in dev)
--------------------------------------------------------------
PointsBuilder::colors        1.784 |  2.060
PointCloudDrawData::new()    1.179 |  0.949
Overall                      9.961 |  9.748

===============================================================

colmap with lots of history

DEV:
                            before   |   after
--------------------------------------------------------------
PointsBuilder::colors         10.160 | 10.486
PointCloudDrawData::new()      6.533 |  4.209
Overall                       46.420 | 43.941

RELEASE:
                            before   |   after
--------------------------------------------------------------
PointsBuilder::colors         13.770 |  6.766
PointCloudDrawData::new()      3.682 |  2.921
Overall                       42.257 | 34.285
```

Btw. there is a deeper rabbit hole we **ignore** here: One could leverage integrated gpus (or the cache of dedicated ones) where the gpu would use the memory directly, eliminating the need to schedule a gpu memcpy, however this is behind a wgpu feature flag anyways (`MAPPABLE_PRIMARY_BUFFERS`).

### Checklist
* [x] I have read and agree to [Contributor Guide](../CONTRIBUTING.md) and the [Code of Conduct](../CODE_OF_CONDUCT.md)
* [ ] I've included a screenshot or gif (if applicable)
* [ ] I've added a line to `CHANGELOG.md` (if this is a big enough change to warrant it)

First half of https://github.com/rerun-io/rerun/issues/426

### CODE REVIEW

File: crates/re_renderer/src/allocator/cpu_write_gpu_read_belt.rs [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+    /// as source.
+    closed_chunks: Vec<Chunk>,
+
+    /// Chunks that are back from the GPU and ready to be mapped for write and put
+    /// into `active_chunks`.
+    free_chunks: Vec<Chunk>,
+
+    /// When closed chunks are mapped again, the map callback sends them here.
+    ///
+    /// Behind a mutex, so that our StagingBelt becomes Sync.
```
Comment (teh-cmc) [inline crates/re_renderer/src/allocator/cpu_write_gpu_read_belt.rs:235]: Not anymore!

File: crates/re_renderer/src/allocator/cpu_write_gpu_read_belt.rs [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+}
+
+impl CpuWriteGpuReadBelt {
+    /// All allocations of this allocator will be aligned to at least this size.
+    ///
+    /// Requiring a minimum alignment means we need to pad less often.
+    /// Also, it has the potential of making memcpy operations faster.
+    ///
+    /// Align to 4xf32. Should be enough for most usecases!
+    pub const MIN_ALIGNMENT: u64 = 16;
```
Comment (teh-cmc) [inline crates/re_renderer/src/allocator/cpu_write_gpu_read_belt.rs:251]: A link to wgpu's alignment sizes table would be nice to help put things into context here.

Comment (teh-cmc) [reply crates/re_renderer/src/allocator/cpu_write_gpu_read_belt.rs:251]: Probably a doc link to `wgpu::MAP_ALIGNMENT` too

File: crates/re_renderer/src/context.rs [hunk 3]
```diff
@@ -146,28 +180,60 @@ impl RenderContext {
...
 
-            frame_index: 0,
+            inflight_queue_submissions: Vec::new(),
+
+            active_frame: ActiveFrameContext { frame_global_command_encoder: None,             per_frame_data_helper: TypeMap::new(),
+                 frame_index: 0 }
+        }
+    }
+
+    fn poll_device(&mut self) {
```
Comment (teh-cmc) [inline crates/re_renderer/src/context.rs:201]: _neat_

File: crates/re_renderer/src/point_cloud_builder.rs [hunk 4]
```diff
@@ -150,13 +153,14 @@ where
...
-                std::iter::repeat(Color32::WHITE).take(self.0.vertices.len() - self.0.colors.len()),
+        if self.0.color_buffer.num_written() < self.0.vertices.len() {
+            self.0.color_buffer.extend(
+                std::iter::repeat(Color32::WHITE)
+                    .take(self.0.vertices.len() - self.0.color_buffer.num_written()),
             );
         }
 
-        if self.0.user_data.len() < self.0.user_data.len() {
+        if self.0.user_data.len() < self.0.vertices.len() {
```
Comment (teh-cmc) [inline crates/re_renderer/src/point_cloud_builder.rs:163]: nice save :trollface:

**Review by teh-cmc — APPROVED** [review]
Marvellous.

File: crates/re_renderer/src/allocator/cpu_write_gpu_read_belt.rs (hunk 1 above)
Comment (Wumpf) [reply crates/re_renderer/src/allocator/cpu_write_gpu_read_belt.rs:235]: these comments didn't age well 😅

File: crates/re_renderer/src/point_cloud_builder.rs (hunk 4 above)
Comment (Wumpf) [reply crates/re_renderer/src/point_cloud_builder.rs:163]: that caused me some reeeaaal weird crash :D

**Pull request merged**